cleo_plugins/FileSystemOperations/FileSystemOperations.cpp

```diff
@@ -315,25 +317,66 @@ class FileSystemOperations
         auto data = tmpBuff.data();
 
         bool ok = File::readString(handle, data, size) != nullptr;
-        if(!ok)
+        if (!ok)
         {
             OPCODE_CONDITION_RESULT(false);
             return OR_CONTINUE;
         }
 
-        // copy into result param
-        int len = strlen(data);
-        int resultSize = min(len, bufferSize - (int)needsTerminator);
+        // In older CLEO this opcode only accepted output as reference to short/long string variables.
+        // If user wanted to load string into memory address, the only way was some variable array with carefully calculated out-of-bouds index,
+        // resulting in indexed dummy element to appear under that address. Detect such cases:
+        bool externalAddress = false;
+        switch (resultParamType)
+        {
+            // global variable array
+            case DT_VAR_ARRAY:
+            case DT_VAR_TEXTLABEL:
+            case DT_VAR_TEXTLABEL_ARRAY:
+            case DT_VAR_STRING_ARRAY:
+            {
+                auto globalsBeg = CTheScripts::ScriptSpace;
+                auto globalsEnd = globalsBeg + 16381 * sizeof(SCRIPT_VAR); // 16381 is count of global variables for GTA 3/VC/SA/SA-Mobile
+                externalAddress = buffer < globalsBeg || buffer > globalsEnd;
+                break;
+            }
+
+            // local variable array
+            case DT_LVAR_ARRAY:
+            case DT_LVAR_TEXTLABEL:
+            case DT_LVAR_TEXTLABEL_ARRAY:
+            case DT_LVAR_STRING_ARRAY:
+            {
+                auto localsBeg = (char*)thread + offsetof(CLEO::CRunningScript, LocalVar);
+                auto localsEnd = localsBeg + sizeof(CLEO::CRunningScript::LocalVar) + sizeof(CLEO::CRunningScript::Timers); // timers are in fact just two extra local variables
```

Comment:
that's not the original behavior. just bring back original implementation from CLEO4.

If you need some safety checks,

Reply:
When user provide long string variable as target he expects to read into it, no into 5 other next variables if file data happens to be longer.
The fact you can write 40 characters into 4@v does not mean any other opcode will be able to read it, as these types are not based on null termination.

Reply:
Maybe it makes sense anyway. That data can be accessed with GetVarPointer. FileRead also does not care about target size.

Reply:
If I provide an int value or int variable, or element of an int array, copy {size} bytes from file to that address.
If I provide a string variable (or element of a string array), copy {size} bytes to the address of that variable.

Don't limit users on what they can and can not do.

Reply:
I may use entire local variables space (32 variables) as a temp buffer to read chunks of data from file (128 bytes), modify them and write back. I don't need any intervention with data from runtime, just raw copy.

Reply:
Only thing we should do is to ensure null-terminator at the {size} position, if fgets does not do it already.
And gate negative size/invalid pointers.

Reply:
Updated.
fgets reads size-1 by default.

Reply:
I'm thinking of a use case like:
```
0@v = "0123456789ABCDEF"
read_string_from_file file_handle to 0@v size 16
```
will the string end with \0?

Reply:
Yes, 15 characters and terminator is written.

Reply:
```
{$CLEO .cs}
nop

0@ = open_file "cleo\.cleo.log" {mode} "r" // tested opcode
1@v = "0123456789ABCDEF"
read_string_from_file 0@ to 1@v size 16

0AD1: print_formatted_now fmt "%s" time 200000 arg 1@v

terminate_this_custom_script
```

Reply:
I mean, can you update test suite?

Reply:
Ehhh, ok :]

Reply:
probably makes sense to gradually built out comprehensive test suite, one fix at a time